### PR TITLE
Close partial indices after merging for batch indexing

### PR DESCRIPTION
**Pull request by navis**
hadoop-index task merges partial indices into one before publishing. But mergee indices are not closed explicitly, adding GC pressure. I can see some more cases in append task, etc. but left those alone.

### CODE REVIEW

File: processing/src/main/java/io/druid/segment/IndexMerger.java [hunk 1]
```diff
@@ -214,7 +214,26 @@ public File mergeQueryableIndex(
...
+      final AggregatorFactory[] metricAggs,
+      File outDir,
+      IndexSpec indexSpec,
+      ProgressIndicator progress
+  ) throws IOException
+  {
+    try {
+      return mergeQueryableIndex(indexes, rollup, metricAggs, outDir, indexSpec, progress);
+    }
+    finally {
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: Maybe use Closer? https://google.github.io/guava/releases/16.0/api/docs/com/google/common/io/Closer.html

Comment (navis) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: My first intension was using that but it made change much bigger. In my experience, it'll make PRs hard to be committed. If any committer grants me, I'll do that.

Comment (gianm) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: the `true` to Closeables.close should ensure we try to close them all (errors will be skipped).

Comment (gianm) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: Hmm, we used to have two versions of `mergeQueryableIndex` that behaved the same, except one let callers specify a progress indicator. With this patch, now the no-progress-indicator version keeps inputs open, but the progress-indicator version closes inputs. It also has a different name. IMO we should resolve this inconsistency before merging this patch: either both methods should close inputs, or neither should.

If both, we should make sure none of the call sites need the indexes to remain open.

If neither, I think it's good to keep the spirit of this patch (closing input indexes during hadoop indexing) by having the call site in IndexGeneratorJob do the closing.

Comment (jon-wei) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: Looks like some call sites such as unit tests call the no-progress-indicator version and close the subindexes themselves, those may need adjusting, if we go with both method closing the subindexes. I think it's fine if that's the approach taken and that makes this PR bigger.

**Pull request closed without merging**